File: packages/sdk-java/harness-sdk-core/src/main/java/com/harness/guardrails/StreamInterceptor.java
package com.harness.guardrails;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * Token-level stream interceptor.
 *
 * Real-time content safety monitoring during streaming output.
 * Detects and interrupts harmful content during generation.
 *
 * Example:
 * <pre>
 * ComplianceJudge judge = new ComplianceJudge(judgeConfig);
 * StreamInterceptor interceptor = new StreamInterceptor(judge, streamConfig);
 *
 * // Check streaming content
 * for (String chunk : stream) {
 *     InterceptResult result = interceptor.checkChunk(chunk).join();
 *     if (result.shouldStop()) {
 *         break;
 *     }
 *     yield chunk;
 * }
 * </pre>
 */
public class StreamInterceptor {

    private static final Logger logger = LoggerFactory.getLogger(StreamInterceptor.class);

    private final ComplianceJudge judge;
    private final GuardrailConfig.StreamInterceptConfig config;
    private final ObjectMapper objectMapper;

    public StreamInterceptor(ComplianceJudge judge, GuardrailConfig.StreamInterceptConfig config) {
        this.judge = judge;
        this.config = config;
        this.objectMapper = new ObjectMapper();

        logger.info(
            "StreamInterceptor initialized: checkInterval={}, safetyThreshold={}",
            config.getCheckInterval(),
            config.getSafetyThreshold()
        );
    }

    /**
     * Intercept result.
     */
    public static class InterceptResult {
        private final boolean shouldStop;
        private final double safetyScore;
        private final String reason;
        private final int tokensChecked;

        public InterceptResult(boolean shouldStop, double safetyScore, String reason, int tokensChecked) {
            this.shouldStop = shouldStop;
            this.safetyScore = safetyScore;
            this.reason = reason;
            this.tokensChecked = tokensChecked;
        }

        public boolean shouldStop() { return shouldStop; }
        public double safetyScore() { return safetyScore; }
        public String reason() { return reason; }
        public int tokensChecked() { return tokensChecked; }
    }

    /**
     * Check a streaming chunk for safety.
     *
     * @param chunk The SSE data line
     * @param buffer Accumulated content buffer
     * @param tokenCount Current token count
     * @return CompletableFuture with InterceptResult
     */
    public CompletableFuture<InterceptResult> checkChunk(String chunk, StringBuilder buffer, int tokenCount) {
        if (!config.isEnabled()) {
            return CompletableFuture.completedFuture(
                new InterceptResult(false, 1.0, "Stream interception disabled", tokenCount)
            );
        }

        // Extract text from SSE
        String text = extractTextFromSse(chunk);
        if (text != null && !text.isEmpty()) {
            buffer.append(text);
        }

        // Check if we should run safety check
        boolean shouldCheck = tokenCount >= config.getMinTokensBeforeCheck() &&
                              tokenCount % config.getCheckInterval() == 0;

        if (!shouldCheck) {
            return CompletableFuture.completedFuture(
                new InterceptResult(false, 1.0, "Not checking yet", tokenCount)
            );
        }

        // Run quick check
        return judge.quickCheck(buffer.toString())
            .thenApply(score -> {
                logger.debug("Stream safety check: tokens={}, score={}", tokenCount, score);

                if (score < config.getSafetyThreshold()) {
                    logger.warn(
                        "Stream interrupted: score={} < threshold={}, tokens={}",
                        score, config.getSafetyThreshold(), tokenCount
                    );
                    return new InterceptResult(true, score, "Content risk detected", tokenCount);
                }

                return new InterceptResult(false, score, "Content safe", tokenCount);
            })
            .exceptionally(e -> {
                logger.warn("Safety check failed: {}, continuing stream", e.getMessage());
                return new InterceptResult(false, 0.5, "Check failed: " + e.getMessage(), tokenCount);
            });
    }

    /**
     * Check content safety (non-streaming).
     *
     * @param content Content to check
     * @return CompletableFuture with InterceptResult
     */
    public CompletableFuture<InterceptResult> checkContent(String content) {
        if (!config.isEnabled()) {
            return CompletableFuture.completedFuture(
                new InterceptResult(false, 1.0, "Stream interception disabled", content.split("\\s+").length)
            );
        }

        return judge.quickCheck(content)
            .thenApply(score -> {
                boolean shouldStop = score < config.getSafetyThreshold();

                return new InterceptResult(
                    shouldStop,
                    score,
                    shouldStop ? "Content risk detected" : "Content safe",
                    content.split("\\s+").length
                );
            })
            .exceptionally(e -> {
                logger.warn("Content check failed: {}", e.getMessage());
                return new InterceptResult(
                    false,
                    0.5,
                    "Check failed: " + e.getMessage(),
                    content.split("\\s+").length
                );
            });
    }

    /**
     * Extract text from SSE line.
     *
     * Supports OpenAI and Claude format SSE data.
     *
     * @param line SSE data line
     * @return Extracted text, or null if not text content
     */
    public String extractTextFromSse(String line) {
        if (line == null || line.isEmpty()) {
            return null;
        }

        line = line.trim();

        // Skip empty lines and non-data lines
        if (!line.startsWith("data: ")) {
            return null;
        }

        String data = line.substring(6);  // Remove "data: " prefix

        // Skip end marker
        if ("[DONE]".equals(data)) {
            return null;
        }

        try {
            JsonNode parsed = objectMapper.readTree(data);

            // OpenAI format
            if (parsed.has("choices")) {
                JsonNode choices = parsed.path("choices");
                if (choices.isArray() && choices.size() > 0) {
                    JsonNode delta = choices.get(0).path("delta");
                    if (delta.has("content")) {
                        return delta.path("content").asText();
                    }
                }
            }

            // Claude format
            if (parsed.has("delta")) {
                JsonNode delta = parsed.path("delta");
                if (delta.has("text")) {
                    return delta.path("text").asText();
                }
            }

            return null;

        } catch (Exception e) {
            return null;
        }
    }

    /**
     * Create interrupt message in SSE format.
     *
     * @param score Safety score
     * @return SSE-formatted interrupt message
     */
    public String createInterruptMessage(double score) {
        String message = String.format(
            "\n\n[内容安全警告：检测到潜在风险内容，输出已中断。安全评分：%.2f]",
            score
        );

        try {
            Map<String, Object> data = Map.of(
                "choices", List.of(Map.of(
                    "delta", Map.of("content", message),
                    "finish_reason", "content_filter",
                    "index", 0
                )),
                "object", "chat.completion.chunk"
            );

            return "data: " + objectMapper.writeValueAsString(data) + "\n\n";
        } catch (Exception e) {
            return "data: {\"choices\":[{\"delta\":{\"content\":\"[Content safety warning]\"},\"finish_reason\":\"content_filter\"}]}\n\n";
        }
    }
}

File: packages/sdk/src/harness/guardrails/stream_interceptor.py
"""
流式拦截器模块

实现 Token 级实时监控和熔断，用于流式响应中的内容安全检测。
"""

import logging
from collections.abc import AsyncIterator
from dataclasses import dataclass

from harness.guardrails.config import StreamInterceptConfig
from harness.guardrails.judge import ComplianceJudge

logger = logging.getLogger(__name__)


@dataclass
class InterceptResult:
    """拦截结果"""

    should_stop: bool
    safety_score: float
    reason: str
    tokens_checked: int


class StreamInterceptor:
    """
    Token 级流式拦截器

    在流式输出过程中实时监控内容，检测到违规内容时立即中断。
    """

    def __init__(self, judge: ComplianceJudge, config: StreamInterceptConfig):
        self.judge = judge
        self.config = config
        logger.info(
            f"StreamInterceptor initialized: check_interval={config.check_interval}, "
            f"safety_threshold={config.safety_threshold}"
        )

    async def intercept(
        self, stream: AsyncIterator[str], context: str | None = None
    ) -> AsyncIterator[str]:
        """
        拦截流式输出，检测有害内容

        Args:
            stream: 原始 SSE 流
            context: 可选的上下文信息

        Yields:
            过滤后的 SSE 数据行

        Raises:
            StreamInterruptException: 当检测到违规内容时
        """
        buffer = ""
        token_count = 0
        check_count = 0

        async for line in stream:
            # 提取文本内容
            text = self._extract_text_from_sse(line)
            if text:
                buffer += text
                token_count += len(text.split())  # 简单的 token 计数

            # 判断是否需要检测
            should_check = (
                token_count >= self.config.min_tokens_before_check
                and token_count % self.config.check_interval == 0
                and check_count < token_count // self.config.check_interval  # 避免重复检测
            )

            if should_check and self.config.enabled:
                check_count = token_count // self.config.check_interval
                try:
                    score = await self.judge.quick_check(buffer)

                    logger.debug(
                        f"Stream safety check #{check_count}: "
                        f"tokens={token_count}, score={score:.2f}"
                    )

                    if score < self.config.safety_threshold:
                        # 检测到风险，中断流
                        logger.warning(
                            f"Stream interrupted: score={score:.2f} < "
                            f"threshold={self.config.safety_threshold}, "
                            f"tokens={token_count}"
                        )
                        # 先发送中断消息
                        yield self._create_interrupt_message(score)
                        return

                except Exception as e:
                    logger.warning(f"Safety check failed: {e}, continuing stream")
                    # 检测失败时继续流（保守策略）

            # 无论是否检测，都转发原始行
            yield line

    async def check_content(self, content: str) -> InterceptResult:
        """
        检查内容安全性（非流式）

        Args:
            content: 待检查的内容

        Returns:
            InterceptResult: 检查结果
        """
        try:
            score = await self.judge.quick_check(content)
            should_stop = score < self.config.safety_threshold

            return InterceptResult(
                should_stop=should_stop,
                safety_score=score,
                reason="Content safety check" if not should_stop else "Content risk detected",
                tokens_checked=len(content.split()),
            )

        except Exception as e:
            logger.warning(f"Content check failed: {e}")
            return InterceptResult(
                should_stop=False,
                safety_score=0.5,
                reason=f"Check failed: {e}",
                tokens_checked=len(content.split()),
            )

    def _extract_text_from_sse(self, line: str) -> str | None:
        """
        从 SSE 行中提取文本内容

        支持 OpenAI 和 Claude 格式的 SSE 数据。
        """
        import json

        line = line.strip()

        # 跳过空行和非数据行
        if not line or not line.startswith("data: "):
            return None

        data = line[6:]  # 去掉 "data: " 前缀

        # 跳过结束标记
        if data == "[DONE]":
            return None

        try:
            parsed = json.loads(data)

            # OpenAI 格式
            if "choices" in parsed:
                for choice in parsed["choices"]:
                    delta = choice.get("delta", {})
                    if "content" in delta:
                        return delta["content"]

            # Claude 格式
            if "delta" in parsed:
                delta = parsed["delta"]
                if "text" in delta:
                    return delta["text"]

            return None

        except json.JSONDecodeError:
            return None

    def _create_interrupt_message(self, score: float) -> str:
        """创建中断消息的 SSE 格式"""
        import json

        message = f"\n\n[内容安全警告：检测到潜在风险内容，输出已中断。安全评分：{score:.2f}]"

        # 构造 OpenAI 格式的 SSE 数据
        data = {
            "choices": [
                {
                    "delta": {"content": message},
                    "finish_reason": "content_filter",
                    "index": 0,
                }
            ],
            "object": "chat.completion.chunk",
        }

        return f"data: {json.dumps(data, ensure_ascii=False)}\n\n"
